---
title: "Megans Deeda RNASeq DESeq Analysis"
author: "Tanner Porter"
date: "2023-12-07"
output:
  pdf_document: default
  html_document: default
editor_options:
  chunk_output_type: console
---



```{r setup, include=FALSE}
#loadpackages
library(tidyverse)
library(DESeq2)
```

# Basic Manipulation of the Data
Import the count table (sequencing data), metadata, pair them and create a DESeq object.

Megan, I would add my commands to generate the feature tables here....

```{r Data Import}

#Read in the sequencing data...
TB_RNA <- as.matrix(read.csv("/scratch/mfolkerts/Deeda_RNA_Seq/Latest_data/countTable_dedup_nocontrols.txt", sep="\t", row.names = "Geneid"))

#Read in the metadata file.
Metadata <- read.csv("/scratch/mfolkerts/Deeda_RNA_Seq/Latest_data/metadata_RNASeq_Deeda.csv", row.names=1)

#Convert metadata to factors
Metadata$Group <- factor(str_replace(Metadata$Group, " ", "_"), levels = c("CONTROL_2", "CONTROL_1", "CASES")) #Remove was from names

Metadata$Run <- factor(Metadata$Run)


Metadata$Simple_Group <- ifelse(grepl("CASES",Metadata$Group), "CASES", "CONTROL")

#Convert metadata to factors
Metadata$Simple_Group <- factor(Metadata$Simple_Group, levels = c("CONTROL", "CASES")) #Remove was from names

Metadata$Simple_Group

#See if there are issues between the dataset.
rownames(Metadata) %in% colnames(TB_RNA)

#reorder the columns in metatdata and sampledata so that they are in the same order for DeSeq2
TB_RNA <- TB_RNA[, rownames(Metadata)]

#check if it worked
all(rownames(Metadata) == colnames(TB_RNA))
```
# Join with annotations file
```{r}
genes.anno <- fst::read_fst("/scratch/tporter/TB_20231117_Megans_DESeq/GRCh38_GENCODE40LNCipedia_geneAnnotations.fst") %>%
  dplyr::select(gene_id, gene_type, gene_name, biotype_class) %>%
  unique()

```

# General Data Exploration and Visualization
```{r}
Metadata <- Metadata %>%
  mutate(Name = row.names(Metadata))

Metadata %>%  
  ggplot(aes(y = Read_count, x = Name)) +
    geom_bar(stat = "identity")+
    theme_bw()+
    scale_y_log10()+
    theme(axis.text.x = element_text(angle = 90, hjust = 1))

Metadata <- full_join(Metadata, data_frame(Name = colnames(TB_RNA),
                   N_genes = TB_RNA %>%colSums()))

TB_RNA_DF <- as_data_frame(TB_RNA)

TB_RNA_DF$Gene <- row.names(TB_RNA)

TB_RNA_DF <- TB_RNA_DF %>%
  pivot_longer(!Gene, values_to = "Reads", names_to = "Name")

TB_RNA_DF %>%
  filter(Reads > 0) %>%
  group_by(Name) %>%
  tally(name = "N_Genes_0") %>%
  ggplot(aes(x = Name, y = N_Genes_0))+
    geom_bar(stat = "identity")+
    theme_bw()+
    theme(axis.text.x = element_text(angle = 90, hjust = 1))

left_join(Metadata,
          TB_RNA_DF %>%
            filter(Reads > 0) %>%
            group_by(Name) %>%
            tally(name = "N_Genes_0")) %>%
  ggplot(aes(x = Read_count, y = N_Genes_0, col = Group)) +
    geom_point()+
    theme_bw()

left_join(Metadata,
          TB_RNA_DF %>%
            filter(Reads > 5) %>%
            group_by(Name) %>%
            tally(name = "N_Genes_5")) %>%
  ggplot(aes(x = Read_count, y = N_Genes_5, col = Group)) +
    geom_point()+
    theme_bw()


TB_RNA_DF %>%
  filter(Reads > 0) %>%
  group_by(Name) %>%
  tally(name = "N_Genes_0") %>%
  summarise(median = median(N_Genes_0, col = Group))

Metadata <- left_join(Metadata,
          TB_RNA_DF %>%
            filter(Reads > 0) %>%
            group_by(Name) %>%
            tally(name = "N_Genes_0")) %>%
            mutate(Gene_Count_Group = ifelse(N_Genes_0 > 25730, "High", "Low"))

```




```{r}
#This creates a DESeq object
DeedaRNASeq <- DESeqDataSetFromMatrix(countData = TB_RNA,
                              colData = Metadata,
                              design = ~ Group + Run)

DeedaRNASeq
```


# Begin DESeq Analysis for simple control groups
```{r}
#This creates a DESeq object
DeedaRNASeq_Simple_Groups <- DESeqDataSetFromMatrix(countData = TB_RNA,
                              colData = Metadata,
                              design = ~ Simple_Group + Run)

DeedaRNASeq_Simple_Groups
```

# Conduct DESeq Analysis for full groups

```{r DESeq Analysis}
#Conduct RNASeq Analysis for full control
DeedaRNASeq_Analysis <- DESeq(DeedaRNASeq)

plotDispEsts(DeedaRNASeq_Analysis)

resultsNames(DeedaRNASeq_Analysis)

#Create results table
SignificantResults <- results(DeedaRNASeq_Analysis, alpha=0.05)

#Summarize Results Table
summary(SignificantResults)
```

# Conduct DESeq Analysis for simple groups

```{r DESeq Analysis}
#Conduct RNASeq Analysis for full control
DeedaRNASeq_Analysis_Simple_Groups <- DESeq(DeedaRNASeq_Simple_Groups)

plotDispEsts(DeedaRNASeq_Analysis_Simple_Groups)
  
resultsNames(DeedaRNASeq_Analysis_Simple_Groups)

#Create results table
SignificantResults_Simple_Groups <- results(DeedaRNASeq_Analysis_Simple_Groups, alpha=0.05)

#Summarize Results Table
summary(SignificantResults_Simple_Groups)
```

# Contrast Analysis
```{r}

resultsNames(DeedaRNASeq_Analysis_Simple_Groups)

DeedaRNASeq_Analysis_Simple_Groups$Simple_Group

results(DeedaRNASeq_Analysis_Simple_Groups, contrast=c("Simple_Group","CONTROL", "CASES"), alpha=0.05)

Contrast_Simple_CONTROL_CASES <- data.frame(results(DeedaRNASeq_Analysis_Simple_Groups, contrast=c("Simple_Group","CONTROL", "CASES"), alpha=0.05)) %>%
  filter(padj < 0.05) %>%
  mutate(Contrast = "Simple- Control vs. Cases") %>%
  mutate(gene_id = row.names(.))

Contrast_Groups_CONTROL1_CASES <- data.frame(results(DeedaRNASeq_Analysis, contrast=c("Group","CONTROL_1","CASES"), alpha=0.05)) %>%
  filter(padj < 0.05) %>%
  mutate(Contrast = "Full- Control_1 vs. Cases")%>%
  mutate(gene_id = row.names(.))

Contrast_Groups_CONTROL2_CASES <- data.frame(results(DeedaRNASeq_Analysis, contrast=c("Group","CONTROL_2","CASES"), alpha=0.05)) %>%
  filter(padj < 0.05) %>%
  mutate(Contrast = "Full- Control_2 vs. Cases")%>%
  mutate(gene_id = row.names(.))

Contrast_Groups_CONTROL2_CONTROL1 <- data.frame(results(DeedaRNASeq_Analysis, contrast=c("Group","CONTROL_2","CONTROL_1"), alpha=0.05)) %>%
  filter(padj < 0.05) %>%
  mutate(Contrast = "Full- Control_2 vs. Control_1")%>%
  mutate(gene_id = row.names(.))


data.frame(results(DeedaRNASeq_Analysis, contrast=c("Group", "CONTROL_1","CONTROL_2"), alpha=0.05)) %>%
  filter(padj < 0.05)

model.matrix(design(DeedaRNASeq_Analysis), colData(DeedaRNASeq_Analysis))

resultsNames(DeedaRNASeq_Analysis)
```

# Results Visualizations

```{r}
Out <- rbind(Contrast_Simple_CONTROL_CASES, Contrast_Groups_CONTROL1_CASES, Contrast_Groups_CONTROL2_CASES, Contrast_Groups_CONTROL2_CONTROL1)

Out <- right_join(genes.anno, Out)

write.csv(Out, "/scratch/tporter/TB_20231117_Megans_DESeq/Deeda_RNA_Seq_Compiled_Contrast_Results.csv")

Out %>% 
  group_by(gene_id) %>%
  tally() %>%
  arrange(desc(n))

Temp <- Out %>%
  ggplot(aes(x = gene_name, y = Contrast, fill = log2FoldChange))+
    geom_tile()+
    facet_grid(~gene_type, scales = "free_x", space = "free_x")+
    theme_bw()+
    scale_fill_gradientn(colours = c(RColorBrewer::brewer.pal(name = "Spectral", n = 11)))+
    theme(axis.text.x = element_text(angle = -90, hjust = 1))+
    labs(y = "Analysis", x = "Genes")

Temp

ggsave("/scratch/tporter/TB_20231117_Megans_DESeq/Sig_Expression_Changes_All.png", width = 120, height = 3, limitsize = F)

plotly::ggplotly(Temp)



Temp <- Out %>%
  filter(log2FoldChange > 1 | log2FoldChange < -1) %>%
  ggplot(aes(x = gene_name, y = Contrast, fill = log2FoldChange))+
    geom_tile()+
    facet_grid(~gene_type, scales = "free_x", space = "free_x")+
    theme_bw()+
    scale_fill_gradientn(colours = c(RColorBrewer::brewer.pal(name = "Spectral", n = 11)))+
    theme(axis.text.x = element_text(angle = -90, hjust = 0))+
    labs(y = "Analysis", x = "Genes (Filtered fold change >1 or <-1)")

Temp

ggsave("/scratch/tporter/TB_20231117_Megans_DESeq/Sig_Expression_Changes_Filtered.png", width = 60, height = 4, limitsize = F)

plotly::ggplotly(Temp)

```

# BioMart
```{r}
require('biomaRt')

mart <- useMart('ENSEMBL_MART_ENSEMBL')
mart <- useDataset('hsapiens_gene_ensembl', mart)

# annotLookup <- getBM(
#   mart = mart,
#   attributes = c(
#     'ensembl_gene_id',
#     'gene_biotype',
#     'external_gene_name',
#     'uniprot_gn_symbol',
#     'uniprot_gn_id'),
#   uniqueRows=TRUE)

annotLookup <- getBM(
  mart = mart,
  attributes = c(
    'ensembl_gene_id',
    #'gene_biotype',
    #'external_gene_name',
    'uniprotswissprot'),
  uniqueRows=TRUE)

annotLookup <- annotLookup %>%
  filter(!uniprotswissprot == "")

Out$ensembl_gene_id_2 <- Out$gene_id

Out <- Out %>%
  separate(ensembl_gene_id_2, into = c("ensembl_gene_id", "dot"), sep = "\\.")

Out_New <- left_join(Out, annotLookup, by = c("ensembl_gene_id"))

#Out_New <- Out_New[,c(17, 1:16)]

write.csv(Out_New, "/scratch/tporter/TB_20231117_Megans_DESeq/Deeda_RNA_Seq_Compiled_Contrast_Results_with_SWISSUNIPROT.csv")

#Check <- 
  
Out_New %>%
  group_by(Contrast, gene_type, uniprotswissprot, gene_id) %>%
  tally() %>%
  filter() %>%
  arrange(desc(n))

write.csv(Check, "/scratch/tporter/TB_20231117_Megans_DESeq/Deeda_RNA_Seq_Compiled_Contrast_Results_with_SWISSUNIPROT_RedundancyCheck.csv")
```






# Export results table for full groups

```{r}
#Create output to filter on alpha = 0.05
outputSignificantReults <- data.frame(SignificantResults)

outputSignificantReults_filter <- outputSignificantReults %>%
  filter(padj <0.05) %>%
  mutate(gene_id = row.names(.))

outputSignificantReults_filter <- right_join(genes.anno, 
                                             outputSignificantReults_filter)


#write.csv(outputSignificantReults_filter, "/scratch/tporter/TB_20231117_Megans_DESeq/outputSignificantReults_filter.csv")

outputSignificantReults_filter %>%
  group_by(gene_type) %>%
  tally() %>% 
  arrange(desc(n))


#How many rows shoudl we have gotten?
sum(SignificantResults$padj < 0.05, na.rm=TRUE)

```

# Export results table for simple groups

```{r}
#Create output to filter on alpha = 0.05
outputSignificantReults_Simple_Groups <- data.frame(SignificantResults_Simple_Groups)

outputSignificantReults_Simple_Groups_filter <- outputSignificantReults_Simple_Groups %>%
  filter(padj <0.05) %>%
  mutate(gene_id = row.names(.))

outputSignificantReults_Simple_Groups_filter <- right_join(genes.anno, 
                                             outputSignificantReults_Simple_Groups_filter)

#write.csv(outputSignificantReults_Simple_Groups_filter, "/scratch/tporter/TB_20231117_Megans_DESeq/outputSignificantReults_Simple_Groups_filter.csv")


outputSignificantReults_Simple_Groups_filter %>%
  group_by(gene_type) %>%
  tally() %>% 
  arrange(desc(n))


#How many rows shoudl we have gotten?
sum(SignificantResults_Simple_Groups$padj < 0.05, na.rm=TRUE)
```

# Quick Visualization of Genes across Groups (Simple and Full)
It does not appear to have large differences across analyses with two control groups vs 1.

```{r}
Temp <- full_join(outputSignificantReults_Simple_Groups_filter %>%
            mutate(Simple_Controls = log2FoldChange) %>%
            select(gene_id, gene_name, gene_type, Simple_Controls),
          outputSignificantReults_filter %>%
            mutate(Full_Controls = log2FoldChange) %>%
            select(gene_id, gene_name,gene_type, Full_Controls)) %>%
  mutate(across(ends_with("Controls"), ~ifelse(is.na(.), 0, .)))

Temp %>%
  pivot_longer(4:5, values_to = "log2FoldChange", names_to = "Analysis") %>%
  ggplot(aes(x = gene_name, y = log2FoldChange, fill = gene_type))+
    geom_bar(stat = "identity")+
    facet_wrap(~Analysis, ncol = 1)+
    theme_bw()+
    theme(axis.text.x = element_text(angle = 90, hjust = 1))+
    ggsci::scale_fill_npg()
```


```{r DESeq Analysis}
lfcShrink(DeedaRNASeq_Analysis, coef="Run_2_vs_1", type="apeglm")

```

# Visualization and Analyses

## By Run Analysis
Let's look specifically at by run

```{r}

run_results <- results(DeedaRNASeq_Analysis, name =  "Run_2_vs_1", alpha= 0.05)
summary(run_results)

par(mfrow=c(1,3), mar=c(4,4,2,1))
plotMA(lfcShrink(DeedaRNASeq_Analysis, coef="Run_2_vs_1", type="apeglm"), main="apeglm")
plotMA(lfcShrink(DeedaRNASeq_Analysis, coef="Run_2_vs_1", type="normal"), main="normal")
plotMA(lfcShrink(DeedaRNASeq_Analysis, coef="Run_2_vs_1", type="ashr"), main="ashr")
```

```{r}
control1_cases_results <- results(DeedaRNASeq_Analysis, name =  "Group_CONTROL_1_vs_CASES", alpha= 0.05)
summary(control1_cases_results)

par(mfrow=c(1,3), mar=c(4,4,2,1))
plotMA(lfcShrink(DeedaRNASeq_Analysis, coef="Group_CONTROL_1_vs_CASES", type="apeglm"), main="apeglm")
plotMA(lfcShrink(DeedaRNASeq_Analysis, coef="Group_CONTROL_1_vs_CASES", type="normal"), main="normal")
plotMA(lfcShrink(DeedaRNASeq_Analysis, coef="Group_CONTROL_1_vs_CASES", type="ashr"), main="ashr")
```

# More Visualizations

```{r}
library("pheatmap")

select <- order(rowMeans(counts(DeedaRNASeq_Analysis, normalized=TRUE)),
                decreasing=TRUE)[1:50]
df <- as.data.frame(colData(DeedaRNASeq_Analysis)[,c("Group","Run")])

ntd <- normTransform(DeedaRNASeq_Analysis)

pheatmap(assay(ntd)[select,], cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=FALSE, annotation_col=df)

vsd <- vst(DeedaRNASeq_Analysis, blind=FALSE)

pheatmap(assay(ntd)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df)

plotPCA(vsd, intgroup=c("Gene_Count_Group"))
plotPCA(vsd, intgroup=c("Group", "Run"))
plotPCA(vsd, intgroup=c("Run"))
plotPCA(vsd, intgroup=c("Group"))

plotPCA(ntd, intgroup=c("Gene_Count_Group"))
plotPCA(ntd, intgroup=c("Group", "Run"))
plotPCA(ntd, intgroup=c("Run"))
plotPCA(ntd, intgroup=c("Group"))
```



```{r}
sampleDists <- dist(t(assay(vsd)))
library("RColorBrewer")
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$Group, vsd$Run, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
```




