---
title: "Megans Deeda RNASeq DESeq Analysis"
author: "Tanner Porter"
date: "2023-12-07"
output:
  pdf_document: default
  html_document: default
editor_options:
  chunk_output_type: console
---



```{r setup, include=FALSE}
#loadpackages
library(tidyverse)
library(DESeq2)
```

# Basic Manipulation of the Data
Import the count table (sequencing data), metadata, pair them and create a DESeq object.

```{r Data Import}

#Read in the sequencing data...
TB_RNA <- as.matrix(read.csv("/scratch/mfolkerts/Deeda_RNA_Seq/Latest_data/countTable_dedup_nocontrols.txt", sep="\t", row.names = "Geneid"))

#Next several lines remove genes that are not matched in the functional analysis
TB_RNA_DF <- as_data_frame(TB_RNA)
TB_RNA_DF$Gene <- row.names(TB_RNA)
TB_RNA_DF <- TB_RNA_DF %>%
  pivot_longer(!Gene, values_to = "Reads", names_to = "Name")
Temp <- TB_RNA_DF %>%
  separate(Gene, into = c("ensembl_gene_id", "dot"), sep = "\\.") %>%
  group_by(ensembl_gene_id, Name) %>%
  summarise(Reads_total = sum(Reads))
Temp <- semi_join(Temp, data.frame(ensembl_gene_id =keys(org.Hs.eg.db,keytype="ENSEMBL")))
Temp <- Temp %>%
  pivot_wider(values_from = Reads_total, names_from = Name)
TB_RNA <- as.matrix(Temp %>%
  column_to_rownames("ensembl_gene_id"))


#Read in the metadata file.
Metadata <- readxl::read_xlsx("/scratch/tporter/TB_20231117_Megans_DESeq/TGEN CASS Shipment 1 Metadata_07-02-2024.xlsx")

Metadata_2 <- read_csv("/scratch/mfolkerts/Deeda_RNA_Seq/Latest_data/metadata_RNASeq_Deeda.csv")

Metadata <- left_join(Metadata, Metadata_2)

rm(Metadata_2)

# row.names(Metadata) <- Metadata$`Participant ID` #Depreciated

Metadata <- Metadata %>%
  mutate(Participant_ID2 = `Participant ID`) %>%
  column_to_rownames(var = "Participant_ID2")

#Convert metadata to factors
Metadata$Group <- factor(str_replace(Metadata$Grouping, " ", "_"), levels = c("CONTROL_2", "CONTROL_1", "CASES")) #Remove was from names

Metadata$Run <- factor(Metadata$Run)

Metadata$Gender <- factor(Metadata$Gender)

Metadata$HIV_Status <- factor(Metadata$`HIV Status`)

Metadata$CASS_Smear <- factor(Metadata$`CASS Smear`)

Metadata$CASS_Culture <- factor(Metadata$`CASS Culture`)

Metadata$Simple_Group <- ifelse(grepl("CASES",Metadata$Group), "CASES", "CONTROL")

#Convert metadata to factors
Metadata$Simple_Group <- factor(Metadata$Simple_Group, levels = c("CONTROL", "CASES")) #Remove was from names

Metadata$Simple_Group


########
########
# Removes 1 sample that was not sequenced.
########
########
Metadata <- Metadata %>%
  filter(row.names(Metadata) %in% colnames(TB_RNA))
########
########


#See if there are issues between the dataset.
rownames(Metadata) %in% colnames(TB_RNA)
#reorder the columns in metatdata and sampledata so that they are in the same order for DeSeq2
TB_RNA <- TB_RNA[, rownames(Metadata)]
#check if it worked
all(rownames(Metadata) == colnames(TB_RNA))


##### Since there is a bunch of metadata missing...
Metadata_CASS_SMEAR <- Metadata %>%
  filter(!is.na(`CASS Smear`))



TB_RNA_CASS_SMEAR <- TB_RNA[,(colnames(TB_RNA) %in% rownames(Metadata_CASS_SMEAR)), drop = F]
#check if it worked
all(rownames(Metadata_CASS_SMEAR) == colnames(TB_RNA_CASS_SMEAR)) 


```
# Join with annotations file
```{r}
genes.anno <- fst::read_fst("/scratch/tporter/TB_20231117_Megans_DESeq/GRCh38_GENCODE40LNCipedia_geneAnnotations.fst") %>%
  dplyr::select(gene_id, gene_type, gene_name, biotype_class) %>%
  unique()

```

# General Data Exploration and Visualization
```{r}
Metadata <- Metadata %>%
  mutate(Name = row.names(Metadata))

Metadata <- full_join(Metadata, data_frame(Name = colnames(TB_RNA),
                   N_genes = TB_RNA %>%colSums()))

TB_RNA_DF <- left_join(TB_RNA_DF, Metadata)

Metadata <- left_join(Metadata,
          TB_RNA_DF %>%
            filter(Reads > 0) %>%
            group_by(Name) %>%
            tally(name = "N_Genes_0")) %>%
            mutate(Gene_Count_Group = ifelse(N_Genes_0 > 25730, "High", "Low"))

Metadata$CASS_outcome <- factor(Metadata$`CASS outcome`, levels = c("Negative", "Positive")) #Remove was from names

Metadata$Smear_Group <- factor(Metadata$Smear, levels = c("Negative", "Positive")) #Remove was from names
```



# Conduct DESeq Analysis for Group
```{r}
#This creates a DESeq object
DeedaRNASeq_Groups <- DESeqDataSetFromMatrix(countData = TB_RNA,
                              colData = Metadata,
                              design = ~ Group + Run + Gender + HIV_Status)

DeedaRNASeq_Groups

#Conduct RNASeq Analysis for full control
DeedaRNASeq_Analysis_Groups <- DESeq(DeedaRNASeq_Groups, test="Wald")

plotDispEsts(DeedaRNASeq_Analysis_Groups)

resultsNames(DeedaRNASeq_Analysis_Groups)

DeedaRNASeq_Analysis_Groups@design
```

# Conduct DESeq Analysis for CASS_outcome
```{r}
#This creates a DESeq object
DeedaRNASeq_CASS_outcome <- DESeqDataSetFromMatrix(countData = TB_RNA,
                              colData = Metadata,
                              design = ~ CASS_outcome + Run + Gender + HIV_Status)

DeedaRNASeq_CASS_outcome

#Conduct RNASeq Analysis for full control
DeedaRNASeq_Analysis_CASS_outcome <- DESeq(DeedaRNASeq_CASS_outcome, test="Wald")

plotDispEsts(DeedaRNASeq_Analysis_CASS_outcome)
  
resultsNames(DeedaRNASeq_Analysis_CASS_outcome)

DeedaRNASeq_Analysis_CASS_outcome@design
```

# Conduct DESeq Analysis for CAAS Smear
```{r}
#This creates a DESeq object
DeedaRNASeq_Smear <- DESeqDataSetFromMatrix(countData = TB_RNA_CASS_SMEAR,
                              colData = Metadata_CASS_SMEAR,
                              design = ~ CASS_Smear + Run + Gender + HIV_Status)

DeedaRNASeq_Smear

#Conduct RNASeq Analysis for full control
DeedaRNASeq_Analysis_Smear <- DESeq(DeedaRNASeq_Smear, test="Wald")

plotDispEsts(DeedaRNASeq_Analysis_Smear)
  
resultsNames(DeedaRNASeq_Analysis_Smear)

DeedaRNASeq_Analysis_Smear@design
```

# Conduct DESeq Analysis for CASS Smear
```{r}
#This creates a DESeq object
DeedaRNASeq_Culture <- DESeqDataSetFromMatrix(countData = TB_RNA_CASS_SMEAR,
                              colData = Metadata_CASS_SMEAR,
                              design = ~ CASS_Culture + Run + Gender + HIV_Status)

DeedaRNASeq_Culture

#Conduct RNASeq Analysis for full control
DeedaRNASeq_Analysis_Culture <- DESeq(DeedaRNASeq_Culture, test="Wald")

plotDispEsts(DeedaRNASeq_Analysis_Culture)
  
resultsNames(DeedaRNASeq_Analysis_Culture)

DeedaRNASeq_Analysis_Culture@design
```

# Functional Analysis Using GAGE
```{r}
# BiocManager::install("gage")
# BiocManager::install("org.Hs.eg.db")
# BiocManager::install("AnnotationDbi")
# BiocManager::install("gageData")
# BiocManager::install("pathview")

library("AnnotationDbi")
library("org.Hs.eg.db")
library("gage")

sessionInfo()

columns(org.Hs.eg.db)

res <- results(DeedaRNASeq_Analysis_Groups, contrast=c("Group","CASES","CONTROL_2"), alpha=0.05)
res = res[order(res$pvalue),]
summary(res)

res

keys(org.Hs.eg.db,keytype="ENSEMBL")

res$symbol = mapIds(org.Hs.eg.db,
                     keys=row.names(res), 
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")

res$entrez = mapIds(org.Hs.eg.db,
                     keys=row.names(res), 
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")

res$name =   mapIds(org.Hs.eg.db,
                     keys=row.names(res), 
                     column="GENENAME",
                     keytype="ENSEMBL",
                     multiVals="first")
head(res, 10)
```

```{r}
library(pathview)
library(gage)
library(gageData)
data(kegg.sets.hs)
data(sigmet.idx.hs)
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]
head(kegg.sets.hs, 3)


foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)

# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs, same.dir=TRUE)

keggres

# Look at both up (greater), down (less), and statatistics.
lapply(keggres, head)

keggres$greater

keggres$less

keggres$stats
```

